File: typescript/docs/examples/data-persistence/context-sync-demo.ts
import { AgentBay, ContextSync, CreateSessionParams, newSyncPolicyWithDefaults } from "../../../src";

/**
 * Context Sync Dual-Mode Example for TypeScript SDK
 * 
 * This example demonstrates the dual-mode context.sync() functionality:
 * 1. Async mode with callback - immediate return, result handled via callback
 * 2. Sync mode with await - waits for completion before returning
 */

async function contextSyncWithCallbackDemo(agentBay: AgentBay): Promise<void> {
  console.log("🔄 Starting context sync with callback demo...");
  
  // Step 1: Create context for persistent storage
  console.log("\n📦 Creating context for persistent storage...");
  const contextResult = await agentBay.context.get("sync-callback-demo", true);
  if (!contextResult.success) {
    throw new Error(`Context creation failed: ${contextResult.errorMessage}`);
  }
  const context = contextResult.context!;
  console.log(`✅ Context created: ${context.id}`);
  
  // Step 2: Create session with context sync
  console.log("\n📦 Creating session with context sync...");
  const syncPolicy = newSyncPolicyWithDefaults();
  const contextSync = new ContextSync(
    context.id,
    "/tmp/sync_data",
    syncPolicy
  );
  
  const params: CreateSessionParams = {
    contextSync: [contextSync]
  };
  const sessionResult = await agentBay.create(params);
  if (!sessionResult.success) {
    throw new Error(`Failed to create session: ${sessionResult.errorMessage}`);
  }
  const session = sessionResult.session!;
  console.log(`✅ Session created: ${session.getSessionId()}`);
  
  // Step 3: Create test data
  console.log("\n💾 Creating test data...");
  await session.command.executeCommand("mkdir -p /tmp/sync_data/test_files");
  
  const testFiles = [
    {
      path: "/tmp/sync_data/test_files/small.txt",
      content: "Small test file content\n".repeat(10)
    },
    {
      path: "/tmp/sync_data/test_files/medium.txt", 
      content: "Medium test file content\n".repeat(100)
    },
    {
      path: "/tmp/sync_data/config.json",
      content: JSON.stringify({
        sync_demo: true,
        created_at: new Date().toISOString(),
        session_id: session.getSessionId()
      }, null, 2)
    }
  ];
  
  let createdFiles = 0;
  for (const file of testFiles) {
    const writeResult = await session.fileSystem.writeFile(file.path, file.content);
    if (writeResult.success) {
      console.log(`✅ Created file: ${file.path}`);
      createdFiles++;
    } else {
      console.log(`❌ Failed to create file ${file.path}: ${writeResult.errorMessage}`);
    }
  }
  
  console.log(`📊 Created ${createdFiles}/${testFiles.length} test files`);
  
  try {
    // Method 1: Async interface with callback
    console.log("\n📞 Calling context.sync() with callback...");
    const syncStartTime = Date.now();
    
    // Use callback mode - function returns immediately
    const syncResult = await session.context.sync(
      undefined, // contextId
      undefined, // path  
      undefined, // mode
      (success: boolean) => {
        console.log("Callback called ==============");
        const callbackTime = Date.now();
        const duration = callbackTime - syncStartTime;
        
        if (success) {
          console.log(`✅ Context sync completed successfully in ${duration}ms`);
        } else {
          console.log(`❌ Context sync completed with failures in ${duration}ms`);
        }
        
        // Delete session in callback
        console.log("🗑️  Deleting session from callback...");
        session.delete(false) // Don't sync again since we already did
          .then(() => {
            console.log("✅ Session deleted successfully from callback");
          })
          .catch((error: any) => {
            console.error("❌ Failed to delete session from callback:", error);
          });
      }
    );
    
    console.log(`📤 Sync initiation result: success=${syncResult.success}, requestId=${syncResult.requestId}`);
    console.log("⏳ Waiting for callback to complete...");
    
    // Wait a bit for the callback to complete
    await new Promise(resolve => setTimeout(resolve, 10000));
    
  } catch (error) {
    console.error("❌ Context sync with callback failed:", error);
    // Clean up session
    try {
      await session.delete(false);
      console.log("✅ Session cleaned up after error");
    } catch (deleteError) {
      console.error("❌ Failed to clean up session:", deleteError);
    }
  }
}

async function contextSyncDemo(agentBay: AgentBay): Promise<void> {
  console.log("🔄 Starting context sync demo...");
  
  // Step 1: Create context for persistent storage
  console.log("\n📦 Creating context for persistent storage...");
  const contextResult = await agentBay.context.get("sync-await-demo", true);
  if (!contextResult.success) {
    throw new Error(`Context creation failed: ${contextResult.errorMessage}`);
  }
  const context = contextResult.context!;
  console.log(`✅ Context created: ${context.id}`);
  
  // Step 2: Create session with context sync
  console.log("\n📦 Creating session with context sync...");
  const syncPolicy = newSyncPolicyWithDefaults();
  const contextSync = new ContextSync(
    context.id,
    "/tmp/sync_data",
    syncPolicy
  );
  
  const params: CreateSessionParams = {
    contextSync: [contextSync]
  };
  const sessionResult = await agentBay.create(params);
  if (!sessionResult.success) {
    throw new Error(`Failed to create session: ${sessionResult.errorMessage}`);
  }
  const session = sessionResult.session!;
  console.log(`✅ Session created: ${session.getSessionId()}`);
  
  // Step 3: Create test data
  console.log("\n💾 Creating test data...");
  await session.command.executeCommand("mkdir -p /tmp/sync_data/test_files");
  
  const testFiles = [
    {
      path: "/tmp/sync_data/test_files/small.txt",
      content: "Small test file content\n".repeat(10)
    },
    {
      path: "/tmp/sync_data/test_files/medium.txt", 
      content: "Medium test file content\n".repeat(100)
    },
    {
      path: "/tmp/sync_data/config.json",
      content: JSON.stringify({
        sync_demo: true,
        created_at: new Date().toISOString(),
        session_id: session.getSessionId()
      }, null, 2)
    }
  ];
  
  let createdFiles = 0;
  for (const file of testFiles) {
    const writeResult = await session.fileSystem.writeFile(file.path, file.content);
    if (writeResult.success) {
      console.log(`✅ Created file: ${file.path}`);
      createdFiles++;
    } else {
      console.log(`❌ Failed to create file ${file.path}: ${writeResult.errorMessage}`);
    }
  }
  
  console.log(`📊 Created ${createdFiles}/${testFiles.length} test files`);
  
  try {
    // Method 2: Sync interface with await
    console.log("\n⏳ Calling context.sync() with await...");
    const syncStartTime = Date.now();
    
    // Use await mode - function waits for completion
    const syncResult = await session.context.sync();
    
    const syncDuration = Date.now() - syncStartTime;
    
    if (syncResult.success) {
      console.log(`✅ Context sync completed successfully in ${syncDuration}ms`);
    } else {
      console.log(`❌ Context sync completed with failures in ${syncDuration}ms`);
    }
    
    console.log(`📤 Sync result: success=${syncResult.success}, requestId=${syncResult.requestId}`);
    
    // Delete session
    console.log("🗑️  Deleting session...");
    await session.delete(false); // Don't sync again since we already did
    console.log("✅ Session deleted successfully");
    
  } catch (error) {
    console.error("❌ Context sync failed:", error);
    // Clean up session
    try {
      await session.delete(false);
      console.log("✅ Session cleaned up after error");
    } catch (deleteError) {
      console.error("❌ Failed to clean up session:", deleteError);
    }
  }
}

async function main(): Promise<void> {
  console.log("🔄 AgentBay Context Sync Dual-Mode Example (TypeScript)");
  
  // Initialize AgentBay client
  const agentBay = new AgentBay();
  
  try {
    // Method 1: Async interface with callback
    console.log("\n" + "=".repeat(60));
    console.log("🔄 Method 1: context_sync_with_callback (Async with callback)");
    console.log("=".repeat(60));
    
    // Start the first demo without waiting for it to complete
    const callbackPromise = contextSyncWithCallbackDemo(agentBay);
    console.log("contextSyncWithCallbackDemo started (not awaited)");
    console.log("=".repeat(60));
    
    // Sleep 3 seconds
    console.log("\n⏳ Sleeping 3 seconds before next demo...");
    await new Promise(resolve => setTimeout(resolve, 3000));
    
    // Method 2: Sync interface with await
    console.log("\n" + "=".repeat(60));
    console.log("🔄 Method 2: context_sync (Sync with await)");
    console.log("=".repeat(60));
    await contextSyncDemo(agentBay); // With await
    
    // Wait for the first demo to complete
    console.log("\n⏳ Waiting for callback demo to complete...");
    await callbackPromise;
    
  } catch (error) {
    console.error("❌ Example execution failed:", error);
  }
  
  console.log("✅ Context sync dual-mode example completed");
}

// Run the example
if (require.main === module) {
  main().catch(console.error);
}

export { contextSyncDemo, contextSyncWithCallbackDemo, main };

File: golang/docs/examples/context_sync_demo/context_sync_demo.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
)

// Context Sync Dual-Mode Example for Go SDK
//
// This example demonstrates the dual-mode context.sync() functionality:
// 1. Async mode with callback - immediate return, result handled via callback
// 2. Sync mode with wait - waits for completion before returning

func contextSyncWithCallbackUploadDemo(agentBay *agentbay.AgentBay) error {
	fmt.Println("🔄 Starting context sync with callback demo...")

	// Step 1: Create context for persistent storage
	fmt.Println("\n📦 Creating context for persistent storage...")
	contextResult, err := agentBay.Context.Get("sync-callback-demo", true)
	if err != nil {
		return fmt.Errorf("context creation failed: %w", err)
	}
	context := contextResult.Context
	fmt.Printf("✅ Context created: %s\n", context.ID)

	// Step 2: Create session with context sync
	fmt.Println("\n📦 Creating session with context sync...")
	syncPolicy := agentbay.NewSyncPolicy()
	contextSync, _ := agentbay.NewContextSync(
		context.ID,
		"/tmp/sync_data",
		syncPolicy,
	)

	params := &agentbay.CreateSessionParams{
		ImageId:     "imgc-0a9mg1h2a96j5zxie",
		ContextSync: []*agentbay.ContextSync{contextSync},
	}
	sessionResult, err := agentBay.Create(params)
	if err != nil {
		return fmt.Errorf("failed to create session: %w", err)
	}
	session := sessionResult.Session
	fmt.Printf("✅ Session created: %s\n", session.GetSessionId())

	// Step 3: Create test data
	fmt.Println("\n💾 Creating test data...")
	_, err = session.Command.ExecuteCommand("mkdir -p /tmp/sync_data/test_files")
	if err != nil {
		return fmt.Errorf("failed to create directory: %w", err)
	}

	testFiles := []struct {
		path    string
		content string
	}{
		{
			path:    "/tmp/sync_data/test_files/small.txt",
			content: generateContent("Small test file content\n", 10),
		},
		{
			path:    "/tmp/sync_data/test_files/medium.txt",
			content: generateContent("Medium test file content\n", 100),
		},
		{
			path: "/tmp/sync_data/config.json",
			content: func() string {
				config := map[string]interface{}{
					"sync_demo":  true,
					"created_at": time.Now().Format(time.RFC3339),
					"session_id": session.GetSessionId(),
				}
				jsonData, _ := json.MarshalIndent(config, "", "  ")
				return string(jsonData)
			}(),
		},
	}

	createdFiles := 0
	for _, file := range testFiles {
		writeResult, err := session.FileSystem.WriteFile(file.path, file.content, "overwrite")
		if err != nil {
			fmt.Printf("❌ Failed to create file %s: %v\n", file.path, err)
		} else if writeResult.Success {
			fmt.Printf("✅ Created file: %s\n", file.path)
			createdFiles++
		} else {
			fmt.Printf("❌ Failed to create file %s\n", file.path)
		}
	}

	fmt.Printf("📊 Created %d/%d test files\n", createdFiles, len(testFiles))

	// Method 1: Async interface with callback
	fmt.Println("\n📞 Calling context.sync() with callback...")
	syncStartTime := time.Now()

	// Use callback mode - function returns immediately
	syncResult, err := session.Context.SyncWithCallback(
		"", "", "upload", // contextId, path, mode
		func(success bool) {
			callbackTime := time.Now()
			duration := callbackTime.Sub(syncStartTime)

			if success {
				fmt.Printf("✅ Context sync completed successfully in %v\n", duration)
			} else {
				fmt.Printf("❌ Context sync completed with failures in %v\n", duration)
			}

			// Delete session in callback
			fmt.Println("🗑️  Deleting session from callback...")
			_, deleteErr := session.Delete(false) // Don't sync again since we already did
			if deleteErr != nil {
				fmt.Printf("❌ Failed to delete session from callback: %v\n", deleteErr)
			} else {
				fmt.Println("✅ Session deleted successfully from callback")
			}
		},
		150, 1500, // maxRetries, retryInterval (milliseconds)
	)

	if err != nil {
		return fmt.Errorf("context sync with callback failed: %w", err)
	}

	fmt.Printf("📤 Sync initiation result: success=%t, requestId=%s\n", syncResult.Success, syncResult.RequestID)
	fmt.Println("⏳ Waiting for callback to complete...")

	// Wait a bit for the callback to complete
	time.Sleep(10 * time.Second)

	return nil
}

func contextSyncWithCallbackDownloadDemo(agentBay *agentbay.AgentBay) error {
	fmt.Println("🔄 Starting context sync with callback download demo...")

	// Step 1: Create context for persistent storage
	// fmt.Println("\n📦 Creating context for persistent storage...")
	// contextResult, err := agentBay.Context.Get("sync-callback-download-demo", true)
	// if err != nil {
	// 	return fmt.Errorf("context creation failed: %w", err)
	// }
	// context := contextResult.Context
	// fmt.Printf("✅ Context created: %s\n", context.ID)

	// Step 2: Create session with context sync
	fmt.Println("\n📦 Creating session with context sync...")
	syncPolicy := agentbay.NewSyncPolicy()
	contextSync, _ := agentbay.NewContextSync(
		"SdkCtx-04bdvpz4zzyyzbo5p",
		"/home/wuying",
		syncPolicy,
	)

	params := &agentbay.CreateSessionParams{
		ImageId:     "imgc-0ab5takfwxakoei4g",
		ContextSync: []*agentbay.ContextSync{contextSync},
	}
	sessionResult, err := agentBay.Create(params)
	if err != nil {
		return fmt.Errorf("failed to create session: %w", err)
	}
	session := sessionResult.Session
	fmt.Printf("✅ Session created: %s\n", session.GetSessionId())

	// Step 3: Create sync directory (no test files needed for download)
	// fmt.Println("\n📁 Creating sync directory...")
	// _, err = session.Command.ExecuteCommand("mkdir -p /tmp/sync_data")
	// if err != nil {
	// 	return fmt.Errorf("failed to create directory: %w", err)
	// }
	// fmt.Println("✅ Sync directory created")

	// Method 1: Async interface with callback (download mode)
	fmt.Println("\n📞 Calling context.sync() with callback (download mode)...")
	syncStartTime := time.Now()

	// Use callback mode - function returns immediately
	syncResult, err := session.Context.SyncWithCallback(
		"", "", "download", // contextId, path, mode
		func(success bool) {
			callbackTime := time.Now()
			duration := callbackTime.Sub(syncStartTime)

			if success {
				fmt.Printf("✅ Context sync download completed successfully in %v\n", duration)
			} else {
				fmt.Printf("❌ Context sync download completed with failures in %v\n", duration)
			}

			// Delete session in callback
			fmt.Println("🗑️  Deleting session from callback...")
			_, deleteErr := session.Delete(false) // Don't sync again since we already did
			if deleteErr != nil {
				fmt.Printf("❌ Failed to delete session from callback: %v\n", deleteErr)
			} else {
				fmt.Println("✅ Session deleted successfully from callback")
			}
		},
		150, 1500, // maxRetries, retryInterval (milliseconds)
	)

	if err != nil {
		return fmt.Errorf("context sync with callback download failed: %w", err)
	}

	fmt.Printf("📤 Sync download initiation result: success=%t, requestId=%s\n", syncResult.Success, syncResult.RequestID)
	fmt.Println("⏳ Waiting for callback to complete...")

	// Wait a bit for the callback to complete
	time.Sleep(10 * time.Second)

	return nil
}

func contextSyncDemo(agentBay *agentbay.AgentBay) error {
	fmt.Println("🔄 Starting context sync demo...")

	// Step 1: Create context for persistent storage
	fmt.Println("\n📦 Creating context for persistent storage...")
	contextResult, err := agentBay.Context.Get("sync-await-demo", true)
	if err != nil {
		return fmt.Errorf("context creation failed: %w", err)
	}
	context := contextResult.Context
	fmt.Printf("✅ Context created: %s\n", context.ID)

	// Step 2: Create session with context sync
	fmt.Println("\n📦 Creating session with context sync...")
	syncPolicy := agentbay.NewSyncPolicy()
	contextSync, _ := agentbay.NewContextSync(
		context.ID,
		"/tmp/sync_data",
		syncPolicy,
	)

	params := &agentbay.CreateSessionParams{
		ImageId:     "imgc-0a9mg1h2a96j5zxie",
		ContextSync: []*agentbay.ContextSync{contextSync},
	}
	sessionResult, err := agentBay.Create(params)
	if err != nil {
		return fmt.Errorf("failed to create session: %w", err)
	}
	session := sessionResult.Session
	fmt.Printf("✅ Session created: %s\n", session.GetSessionId())

	// Step 3: Create test data
	fmt.Println("\n💾 Creating test data...")
	_, err = session.Command.ExecuteCommand("mkdir -p /tmp/sync_data/test_files")
	if err != nil {
		return fmt.Errorf("failed to create directory: %w", err)
	}

	testFiles := []struct {
		path    string
		content string
	}{
		{
			path:    "/tmp/sync_data/test_files/small.txt",
			content: generateContent("Small test file content\n", 10),
		},
		{
			path:    "/tmp/sync_data/test_files/medium.txt",
			content: generateContent("Medium test file content\n", 100),
		},
		{
			path: "/tmp/sync_data/config.json",
			content: func() string {
				config := map[string]interface{}{
					"sync_demo":  true,
					"created_at": time.Now().Format(time.RFC3339),
					"session_id": session.GetSessionId(),
				}
				jsonData, _ := json.MarshalIndent(config, "", "  ")
				return string(jsonData)
			}(),
		},
	}

	createdFiles := 0
	for _, file := range testFiles {
		writeResult, err := session.FileSystem.WriteFile(file.path, file.content, "overwrite")
		if err != nil {
			fmt.Printf("❌ Failed to create file %s: %v\n", file.path, err)
		} else if writeResult.Success {
			fmt.Printf("✅ Created file: %s\n", file.path)
			createdFiles++
		} else {
			fmt.Printf("❌ Failed to create file %s\n", file.path)
		}
	}

	fmt.Printf("📊 Created %d/%d test files\n", createdFiles, len(testFiles))

	// Method 2: Sync interface with wait
	fmt.Println("\n⏳ Calling context.sync() with wait...")
	syncStartTime := time.Now()

	// Use wait mode - function waits for completion
	syncResult, err := session.Context.SyncWithCallback("", "", "upload", nil, 150, 1500)
	if err != nil {
		return fmt.Errorf("context sync failed: %w", err)
	}

	syncDuration := time.Since(syncStartTime)

	if syncResult.Success {
		fmt.Printf("✅ Context sync completed successfully in %v\n", syncDuration)
	} else {
		fmt.Printf("❌ Context sync completed with failures in %v\n", syncDuration)
	}

	fmt.Printf("📤 Sync result: success=%t, requestId=%s\n", syncResult.Success, syncResult.RequestID)

	// Delete session
	fmt.Println("🗑️  Deleting session...")
	_, err = session.Delete(false) // Don't sync again since we already did
	if err != nil {
		return fmt.Errorf("failed to delete session: %w", err)
	}
	fmt.Println("✅ Session deleted successfully")

	return nil
}

func main() {
	fmt.Println("🔄 AgentBay Context Sync Dual-Mode Example (Go)")

	// Initialize AgentBay client
	agentBay, err := agentbay.NewAgentBay("")
	if err != nil {
		log.Fatalf("Failed to initialize AgentBay: %v", err)
	}

	// Method 1: Async interface with callback (upload)
	fmt.Println("\n" + "=" + repeat("=", 59))
	fmt.Println("🔄 Method 1: context_sync_with_callback_upload (Async with callback)")
	fmt.Println("=" + repeat("=", 59))

	// Start the first demo without waiting for it to complete
	go func() {
		if err := contextSyncWithCallbackUploadDemo(agentBay); err != nil {
			fmt.Printf("❌ Context sync with callback upload demo failed: %v\n", err)
		}
		fmt.Println("contextSyncWithCallbackUploadDemo completed")
		fmt.Println("=" + repeat("=", 59))
	}()

	// Sleep 3 seconds
	fmt.Println("\n⏳ Sleeping 3 seconds before next demo...")
	time.Sleep(3 * time.Second)

	// Method 2: Async interface with callback (download)
	fmt.Println("\n" + "=" + repeat("=", 59))
	fmt.Println("🔄 Method 2: context_sync_with_callback_download (Async with callback)")
	fmt.Println("=" + repeat("=", 59))

	// Start the second demo without waiting for it to complete
	go func() {
		if err := contextSyncWithCallbackDownloadDemo(agentBay); err != nil {
			fmt.Printf("❌ Context sync with callback download demo failed: %v\n", err)
		}
		fmt.Println("contextSyncWithCallbackDownloadDemo completed")
		fmt.Println("=" + repeat("=", 59))
	}()

	// Sleep 3 seconds
	fmt.Println("\n⏳ Sleeping 3 seconds before next demo...")
	time.Sleep(3 * time.Second)

	// Method 3: Sync interface with wait
	fmt.Println("\n" + "=" + repeat("=", 59))
	fmt.Println("🔄 Method 3: context_sync (Sync with wait)")
	fmt.Println("=" + repeat("=", 59))

	if err := contextSyncDemo(agentBay); err != nil {
		fmt.Printf("❌ Context sync demo failed: %v\n", err)
	}

	// Wait for the async demos to complete
	fmt.Println("\n⏳ Waiting for async demos to complete...")
	time.Sleep(15 * time.Second)

	fmt.Println("✅ Context sync dual-mode example completed")
}

// Helper functions
func generateContent(text string, count int) string {
	result := ""
	for i := 0; i < count; i++ {
		result += text
	}
	return result
}

func repeat(s string, count int) string {
	result := ""
	for i := 0; i < count; i++ {
		result += s
	}
	return result
}
